pygmt/src/grdhisteq.py

```diff
@@ -108,34 +108,36 @@ def _grdhisteq(grid, output_type, **kwargs):
         -------
         :func:`pygmt.grd2cpt`
         """
+        if caller not in ["compute_bins", "equalize_grid"]:
+            raise GMTInvalidInput(f"Unrecognized caller: {caller}.")
 
         with Session() as lib:
-            file_context = lib.virtualfile_from_data(check_kind="raster", data=grid)
-            with file_context as infile:
+            with lib.virtualfile_from_data(
+                check_kind="raster", data=grid
+            ) as vingrid, lib.virtualfile_to_data(
+                kind="dataset", fname=kwargs.get("D")
+            ) as vouttbl:
+                if caller == "compute_bins":
+                    kwargs["D"] = vouttbl
                 lib.call_module(
-                    module="grdhisteq", args=build_arg_string(kwargs, infile=infile)
+                    module="grdhisteq", args=build_arg_string(kwargs, infile=vingrid)
                 )
 
-        if output_type == "file":
-            return None
-        if output_type == "xarray":
-            return load_dataarray(kwargs["G"])
-
-        result = pd.read_csv(
-            filepath_or_buffer=kwargs["D"],
-            sep="\t",
-            header=None,
-            names=["start", "stop", "bin_id"],
-            dtype={
-                "start": np.float32,
-                "stop": np.float32,
-                "bin_id": np.uint32,
-            },
-        )
-        if output_type == "numpy":
-            return result.to_numpy()
+            if output_type == "xarray":
+                return load_dataarray(kwargs["G"])
 
-        return result.set_index("bin_id")
+            result = return_table(
+                session=lib,
+                output_type=output_type,
+                vfile=vouttbl,
+                colnames=["start", "stop", "bin_id"],
+            )
+            if output_type == "pandas":
+                result = result.astype(
+                    {"start": np.float32, "stop": np.float32, "bin_id": np.uint32}
+                )
+                return result.set_index("bin_id")
+            return result
```

Comment:
For this module, we also set the dtypes for the three columns. Maybe this should also be done in the `return_table` function, i.e., change the `return_table` function definition to
```
return_table(session, output_type, vfile, colnames, dtypes=None, index_col=None)
```
then we can call it this way:
```
result = return_table(
    session=lib,
    output_type=output_type,
    vfile=vouttbl,
    colnames=["start", "stop", "bin_id"],
    dtype=[np.float32, np.float32, np.uint32], # dtype=None means the default float64
    index_col="bin_id",  # similar to pd.read_csv's `index_col` parameter
)
```